File: lawAiIOS/lawAI/views/RegisterView.swift
//
//  RegisterView.swift
//  lawAI
//
//  Created by Abdullah on 26/05/2023.
//

import SwiftUI

struct RegisterView: View {
    @State private var firstName = ""
    @State private var lastName = ""
    @State private var email = ""
    @State private var phoneNumber = ""
    @State private var password = ""
    @State private var passwordAgain = ""
    @State private var showAlert = false
    @State private var alertMessage = ""
    @Environment(\.presentationMode) var presentationMode: Binding<PresentationMode>
    
    var body: some View {
        NavigationView {
            VStack {
                // First name
                TextField("First Name", text: $firstName)
                    .textFieldStyle(RoundedBorderTextFieldStyle())
                    .autocapitalization(.none)
                    .onReceive(firstName.publisher.collect()) {
                        self.firstName = String($0.prefix(50))
                    }
                
                // Last name
                TextField("Last Name", text: $lastName)
                    .textFieldStyle(RoundedBorderTextFieldStyle())
                    .autocapitalization(.none)
                    .onReceive(lastName.publisher.collect()) {
                        self.lastName = String($0.prefix(50))
                    }
                
                // Email
                TextField("Email", text: $email)
                    .textFieldStyle(RoundedBorderTextFieldStyle())
                    .autocapitalization(.none)
                    .keyboardType(.emailAddress)
                
                // Phone number
                TextField("Phone Number", text: $phoneNumber)
                    .textFieldStyle(RoundedBorderTextFieldStyle())
                    .keyboardType(.phonePad)
                
                // Password
                SecureField("Password", text: $password)
                    .textFieldStyle(RoundedBorderTextFieldStyle())
                
                // Password Again
                SecureField("Re-enter Password", text: $passwordAgain)
                    .textFieldStyle(RoundedBorderTextFieldStyle())
                
                // Register button
                Button(action: {
                    // Check if fields are valid
                    if isValidFields() {
                        // Handle register action
                        self.presentationMode.wrappedValue.dismiss()
                    } else {
                        self.showAlert = true
                    }
                }) {
                    Text("Register")
                        .fontWeight(.bold)
                        .foregroundColor(.white)
                        .padding()
                        .background(Color.green)
                        .cornerRadius(10)
                }
                .alert(isPresented: $showAlert) {
                    Alert(title: Text("Invalid Input"), message: Text(self.alertMessage), dismissButton: .default(Text("OK")))
                }
            }
            .padding()
            .navigationTitle("Register")
        }
    }
    
    func isValidFields() -> Bool {
        // Check if name is not empty
        if firstName.isEmpty || lastName.isEmpty {
            alertMessage = "Please enter your full name."
            return false
        }
        
        // Check if email is valid
        let emailRegEx = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}"
        let emailPred = NSPredicate(format:"SELF MATCHES %@", emailRegEx)
        if !emailPred.evaluate(with: email) {
            alertMessage = "Please enter a valid email."
            return false
        }
        
        // Check if phone number is valid
        let phoneNumberRegEx = "^05[0-9]{8}$"
        let phoneNumberPred = NSPredicate(format:"SELF MATCHES %@", phoneNumberRegEx)
        if !phoneNumberPred.evaluate(with: phoneNumber) {
            alertMessage = "Please enter a valid phone number."
            return false
        }
        
        // Check if password is not empty and matches
        if password.isEmpty || password != passwordAgain {
            alertMessage = "Please enter a matching password."
            return false
        }
        
        return true
    }
}

    

File: estsheerniFrontend/src/screens/RegisterView.js
import React, { useState } from 'react';
import { View, Text, TextInput, TouchableOpacity, Alert, StyleSheet } from 'react-native';
import AuthenticationManager from './services/AuthenticationManager';
import LanguageContext from './services/LanguageContext';

const RegisterView = () => {
  const { language, handleLanguageChange } = useContext(LanguageContext);
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");
  const [phoneNumber, setPhoneNumber] = useState("");
  const [password, setPassword] = useState("");
  const [passwordAgain, setPasswordAgain] = useState("");

  const handleRegister = async () => {
    if (isValidFields()) {
      try {
        await AuthenticationManager.register(email, password, firstName, lastName, phoneNumber);
        // Registration successful, handle navigation or presentation logic here
        Alert.alert("Registration Success", "You have successfully registered.");
      } catch (error) {
        // Registration failed, handle the error
        Alert.alert("Registration Error", error.message);
      }
    } else {
      Alert.alert("Registration Error", "Please fill in all the fields correctly.");
    }
  };

  const isValidFields = () => {
    // Check if all required fields are not empty
    if (
      firstName.trim() === "" ||
      lastName.trim() === "" ||
      email.trim() === "" ||
      phoneNumber.trim() === "" ||
      password.trim() === "" ||
      passwordAgain.trim() === ""
    ) {
      return false;
    }

    // Check if email is valid
    const emailRegEx = /^[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$/;
    if (!emailRegEx.test(email)) {
      return false;
    }

    // Check if phone number is valid
    const phoneNumberRegEx = /^05[0-9]{8}$/;
    if (!phoneNumberRegEx.test(phoneNumber)) {
      return false;
    }

    // Check if passwords match
    if (password !== passwordAgain) {
      return false;
    }

    return true;
  };


  return (
    <View style={styles.container}>
      <TouchableOpacity onPress={handleLanguageChange} style={styles.languageButton}>
        <Text>{language}</Text>
      </TouchableOpacity>
      <Text style={styles.title}>Register</Text>
      <TextInput
        style={styles.input}
        placeholder="First Name"
        onChangeText={setFirstName}
        value={firstName}
      />
      <TextInput
        style={styles.input}
        placeholder="Last Name"
        onChangeText={setLastName}
        value={lastName}
      />
      <TextInput
        style={styles.input}
        placeholder="Email"
        onChangeText={setEmail}
        value={email}
        keyboardType="email-address"
      />
      <TextInput
        style={styles.input}
        placeholder="Phone Number"
        onChangeText={setPhoneNumber}
        value={phoneNumber}
        keyboardType="numeric"
      />
      <TextInput
        style={styles.input}
        placeholder="Password"
        secureTextEntry
        onChangeText={setPassword}
        value={password}
      />
      <TextInput
        style={styles.input}
        placeholder="Password Again"
        secureTextEntry
        onChangeText={setPasswordAgain}
        value={passwordAgain}
      />
      <TouchableOpacity onPress={handleRegister} style={styles.button}>
        <Text style={styles.buttonText}>Register</Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20,
  },
  languageButton: {
    position: 'absolute',
    top: 20,
    left: 20,
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
    marginBottom: 20,
  },
  input: {
    width: '100%',
    height: 40,
    marginBottom: 10,
    borderColor: 'gray',
    borderWidth: 1,
    paddingHorizontal: 10,
  },
  button: {
    backgroundColor: 'blue',
    paddingHorizontal: 20,
    paddingVertical: 10,
    borderRadius: 5,
  },
  buttonText: {
    color: 'white',
    fontSize: 16,
    fontWeight: 'bold',
  },
});

export default RegisterView;


